SwiftCompilerSources/Sources/SIL/Utilities/WalkUtils.swift

```diff
@@ -370,7 +370,28 @@ extension ValueDefUseWalker {
         // We need to ignore this because otherwise the path wouldn't contain the right `existential` field kind.
         return leafUse(value: operand, path: path)
       }
-      return walkDownUses(ofValue: urc, path: path)
+      // The `unchecked_ref_cast` is designed to be able to cast between
+      // `Optional<ClassType>` and `ClassType`. We need to handle these
+      // cases by checking if the type is optional and adjust the path
+      // accordingly.
+      switch (urc.type.isOptional, urc.fromInstance.type.isOptional) {
+        case (true, false):
+          if walkDownUses(ofValue: urc, path: path.push(.enumCase, index: 0)) == .abortWalk {
+            return .abortWalk
+          }
+          return walkDownUses(ofValue: urc, path: path.push(.enumCase, index: 1))
+        case (false, true):
+          if let path = path.popIfMatches(.enumCase, index: 0) {
+            if walkDownUses(ofValue: urc, path: path) == .abortWalk {
+              return .abortWalk
+            } else if let path = path.popIfMatches(.enumCase, index: 1) {
+              return walkDownUses(ofValue: urc, path: path)
+            }
+          }
```

Comment:
In the path doesn't match the enum case you should call `unmatchedPath` (see the `UncheckedEnumDataInst` case)